src/toxcore/dht_new/dht_impl.rs

```diff
@@ -0,0 +1,298 @@
+    cache: HashMap<HashKeys, PrecomputedKey>,
+}
+
+impl PrecomputedKeys {
+    /// manage hash table for precomputed keys
+    pub fn new () -> PrecomputedKeys {
+        PrecomputedKeys {
+            cache: HashMap::new(),
+        }
+    }
+
+    /// Get precomputed keys
+    /// If the Key is not found in cache, create symmetric key and insert it into cache for later use.
+    pub fn get_symmetric_key (&mut self, key_pair: (&SecretKey, &PublicKey)) -> Result<PrecomputedKey, Error> {
+        let key = HashKeys(key_pair.0.clone(), key_pair.1.clone());
+        match self.cache.get(&key) {       // if symmetric key exists in cache, returns with the value
+            Some(k) => {
+                return Ok(k.clone()); 
+            },
+            None => {},
+        };
+
+        // Key don't exist in cache, So create one
+        // must separate logic into two blocks because self.cache is barrowed mutably
+        let shared_secret = encrypt_precompute(key_pair.1, key_pair.0);
+        self.cache.insert (key, shared_secret.clone());
+        Ok(shared_secret)
+    }
+}
+
+impl DhtPacket {
+    /// create new DhtPacket object
+    pub fn new(shared_secret: &PrecomputedKey, pk: &PublicKey, dp: DhtPacketPayload) -> DhtPacket {
+        let nonce = &gen_nonce();
+        let mut buf = [0; MAX_DHT_PACKET_SIZE];
+        let (_, size) = dp.to_bytes((&mut buf, 0)).unwrap();
+        let payload = seal_precomputed(&buf[..size] , nonce, shared_secret);
+
+        DhtPacket {
+            packet_kind: dp.kind(),
+            pk: *pk,
+            nonce: *nonce,
+            payload: payload,
+        }
+
+    }
+
+    /**
+    Decrypt payload and try to parse it as packet type.
+
+    To get info about it's packet type use
+    [`.kind()`](./struct.DhtPacket.html#method.kind) method.
+
+    Returns `None` in case of faliure:
+
+    - fails to decrypt
+    - fails to parse as given packet type
+    */
+    /* TODO: perhaps switch to using precomputed symmetric key?
+              - given that computing shared key is apparently the most
+                costly operation when it comes to crypto, using precomputed
+                key might (would significantly?) lower resource usage
+
+                Alternatively, another method `get_payloadnm()` which would use
+                symmetric key.
+    */
+    pub fn get_payload(&self, own_secret_key: &SecretKey) -> Result<Option<DhtPacketPayload>, Error>
+    {
+        debug!(target: "DhtPacket", "Getting packet data from DhtPacket.");
+        trace!(target: "DhtPacket", "With DhtPacket: {:?}", self);
+        let decrypted = open(&self.payload, &self.nonce, &self.pk,
+                            own_secret_key)
+            .and_then(|d| Ok(d))
+            .map_err(|e| {
+                debug!("Decrypting DhtPacket failed!");
+                Error::new(ErrorKind::Other,
+                    format!("DhtPacket decrypt error: {:?}", e))
+            });
+
+        match self.packet_kind {
```

Comment:
I propose switch to DhtPacketPayload::from_bytes when #11 is merged - it will reduce code size.